src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,437 @@
+                {
+                    CancellationTokenSource.Token.ThrowIfCancellationRequested();
+                    (BlittableJsonReaderObject metadata, string id, string changeVector) = BatchFromServer.GetMetadataFromBlittable(doc.Data);
+                    lastReceivedChangeVector = changeVector;
+                    SubscriptionConnection.WriteDocumentOrException(context, writer, document: null, doc.Data, metadata, doc.Exception, id, null, null, null);
+                    docsToFlush++;
+
+                    if (await SubscriptionConnection.FlushBatchIfNeededAsync(sendingCurrentBatchStopwatch, SubscriptionId, writer, buffer, TcpConnection, stats: null, _logger, docsToFlush, CancellationTokenSource.Token) == false)
+                        continue;
+
+                    docsToFlush = 0;
+                    sendingCurrentBatchStopwatch.Restart();
+                }
+
+                DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Egor, DevelopmentHelper.Severity.Major, "https://issues.hibernatingrhinos.com/issue/RavenDB-16279");
+                SubscriptionConnection.WriteEndOfBatch(writer);
+
+                AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, $"Flushing docs collected from shard '{shard}'"));
+                await SubscriptionConnection.FlushDocsToClientAsync(SubscriptionId, writer, buffer, TcpConnection, stats: null, _logger, docsToFlush, endOfBatch: true, CancellationTokenSource.Token);
+            }
+
+            batch.LastSentChangeVectorInBatch = lastReceivedChangeVector;
+            await WaitForClientAck(replyFromClientTask);
+            AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, $"Shard '{shard}' got ack from client '{TcpConnection.TcpClient.Client.RemoteEndPoint}'."));
+        }
+
+        internal async Task SendConfirmToClient()
+        {
+            AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, $"Shard subscription connection '{SubscriptionId}' send confirm to client '{TcpConnection.TcpClient.Client.RemoteEndPoint}'."));
+            await WriteJsonAsync(new DynamicJsonValue
+            {
+                [nameof(SubscriptionConnectionServerMessage.Type)] = nameof(SubscriptionConnectionServerMessage.MessageType.Confirm)
+            });
+        }
+
+        public override void Dispose()
+        {
+            if (_isDisposed)
+                return;
+
+            _isDisposed = true;
+
+            base.Dispose();
+
+            foreach (var worker in _shardWorkers)
+            {
+                worker.Value.Dispose();
+            }
+        }
+
+        private class SubscriptionShardHolder : IDisposable
+        {
+            public ShardedSubscriptionWorker Worker;
+            public Task PullingTask;
+            public DateTime? LastErrorDateTime;
+            public RequestExecutor RequestExecutor;
+
+            public void Dispose()
+            {
+                Worker.Dispose();
```

Comment:
can worker be null? Create a proper ctor maybe?

Reply:
added ctor